modules/cluster_estimation/cluster_estimation.py

```diff
@@ -0,0 +1,304 @@
+covariance of each landing pad estimation.
+"""
+
+import sklearn
+import sklearn.datasets
+import sklearn.mixture
+import numpy as np
+
+
+from ..object_in_world import ObjectInWorld
+from ..detection_in_world import DetectionInWorld
+
+
+class ClusterEstimation:
+    """
+    Estimate landing pad locations based on bounding box coordinates
+
+    ATTRIBUTES
+    ----------
+    components: int, default = 10
+        input number of clusters for model
+    
+    rand_state: int, default = 0
+        seed for randomizer, to get consistent results
+    
+    min_activation_threshold: int, default = 100
+        minimum total data points before model can be ran 
+
+    new_points_per_run: int, default = 10
+        minimum number of new data points that must be collected
+        before running model 
+
+    METHODS
+    -------
+    __init__()
+        Sets cluster model run conditions and random state
+
+    clear_all_data()
+        Clears all past accumulated data
+
+    reset_model()
+        unfits the model of any data
+
+    run()
+        Take in list of landing pad detections and return list of estimated landing pad locations
+        if number of detections is sufficient, or if manually forced to run.
+
+    __decide_to_run()
+        Decide when to run cluster estimation model
+
+    __convert_detections_to_point()
+        Convert DetectionInWorld input object to a [x,y] position to store
+    
+    __sort_by_weights()
+        Sort input model output list by weights in descending order
+
+    __get_distance()
+        Calculates distance between a point and a cluster center
+    """
+    # Vgmm Model Hyperparams 
+    __MODEL_INIT_PARAM = "k-means++"
+    __COVAR_TYPE = "spherical"
+    __WEIGHT_CONC_PRIOR = 100
+    __MEAN_PRECISION_PRIOR = 1E-6
+    __MAX_MODEL_ITERATIONS = 1000
+
+    # Hyperparams to clean up model outputs
+    __WEIGHT_DROP_THRESHOLD = 0.1
+    __MAX_COVARIANCE_THRESHOLD = 10
+
+    # Real-world scenario hyperparams
+    __MAX_NUM_COMPONENTS = 10  # assumed maximum number of real landing pads
+
+    def __init__(self, min_activation_threshold=100, min_new_points_to_run=10, random_state = 0):
+
+        # Initalizes VGMM model
+        self.__vgmm_model = sklearn.mixture.BayesianGaussianMixture(
+            covariance_type = self.__COVAR_TYPE,
+            n_components = self.__MAX_NUM_COMPONENTS,
+            random_state = random_state,
+            weight_concentration_prior = self.__WEIGHT_CONC_PRIOR,
+            init_params = self.__MODEL_INIT_PARAM,
+            mean_precision_prior = self.__MEAN_PRECISION_PRIOR,
+            max_iter=self.__MAX_MODEL_ITERATIONS)
+
+        # Points storage
+        self.__all_points = []
+        self.__current_bucket = []
+
+        # Requirements to decide to run
+        self.__min_activation_threshold = min_activation_threshold
+        self.__min_new_points_to_run = min_new_points_to_run
+        self.__has_ran_once = False
+    
+    def clear_all_data(self):
```

Comment:
Why does this exist?

Reply:
This is just for ease of testing, it is more easier to test behaviour at different data amounts on the same worker with a clear all data function